design/api-design.md

```diff
@@ -0,0 +1,37 @@
+# API design
+
+## Requirements
+
+* API should be relatively easy to call (i.e., through a single GET/POST request)
+* It should be reasonably responsive by default (<5s)
+* API returns JSON object
+* If the schema is expected to change a lot, versioned API routes would be preferable.
+
+## Querying the API
+
+The API call includes all the results to be included in the map. This helps facilitate scheduled updates and reduces the need for repeat querying. An example query URL would be `api.epiverse.org/map` (versioned could look like `api.epiverse.org/v1/map`).
+
+## Response
+
+The response for the API is one JSON object containing an array of package names, the relevant metadata, and their respective embedding scores.
+
+```json
+{
```

Comment:
Yes to versioning! Would this be a generic version or would we have anything else included here?

If we plan on versioning the API routes, I figure it will still be helpful as a check to ensure the endpoint is returning the proper version (e.g., `/v1/map` returns a `version: 1`).

Reply:
My understanding is that versioning of the API route would happen if the schema changes.

For changes in the back-end without changes to the schema, I would just include the version number in the response, without the need to keep the API route version in sync.

> Would this be a generic version or would we have anything else included here?

Do you have anything specific in mind when you say "anything else"?

Note that this versioning suggestion applies also to the search engine. Possibly even more so to illustrate the output is indeed deterministic **for a given version of the backend**.

Reply:
Fair enough 😊 To see whether I understood correctly, we will have the following in the end?

1. versioned API routes for the schema (for example `/v1/map`)
2. Model version in the response (for example `modelVersion: 0.0.1`)

If I understood correctly, this is helpful as we will be able to disambiguate new models on the same response schema 👍 

---

The "anything else" I mentioned is unclear to me right now, so not too important I expect. Working towards this statement, I may have meant metadata about the model that would prove relevant. This may be a bit much at this time, if we already have a version number of the model.

Reply:
```suggestion
{
  modelVersion: "0.0.1",
```